web/src/App.tsx

```diff
@@ -1,16 +1,43 @@
-import React from "react";
+import React, { useReducer, useContext, useState } from "react";
 import "./App.css";
 import todos from "./utils/data";
 import { TodoType } from "./utils/types";
 
+type AAState = {
+  todos: TodoType[];
+};
+
+const initialState: AAState = {
+  todos,
+};
+
+type reducerAction = {
+  type: "markTodoComplete" | "markTodoIncomplete";
+  payload: TodoType;
+};
+
+const reducer = (state: AAState, action: reducerAction) => {
+  const todo = action.payload;
+  switch (action.type) {
+    case "markTodoComplete":
+      return {
+        ...state,
+        todos: [
+          state.todos.filter((td) => td.id !== todo.id),
+          { ...todo, completed: true },
+        ],
+      };
+  }
+};
```

Comment:
This looks good.  This is something you will use very often, so it's worth really gaining an understanding of how it's used.  
- [ ] Add the `markTodoIncomplete` case to the reducer .

Here are some resources for using the spread operator.  

Official docs: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax

Practical use article: https://medium.com/coding-at-dawn/how-to-use-the-spread-operator-in-javascript-b9e4a8b06fab